eth-storage-trie/test/resolver.spec.js

```diff
@@ -1,124 +1,94 @@
-        expect(err).to.not.exist()
-        let trieNode = result.value
-        expect(result.remainderPath).to.eql('c/0/a/0/0/')
-        expect(isExternalLink(trieNode)).to.eql(true)
-        done()
-      })
+    it('root node resolves to neck', () => {
+      const rootNode = dagNodes[0]
+      const result = resolver.resolve(rootNode, '0/0/0/c/0/a/0/0/')
+      const trieNode = result.value
+      expect(result.remainderPath).to.eql('c/0/a/0/0')
+      expect(CID.isCID(trieNode)).to.be.true()
     })
 
-    it('neck node resolves "c" down to buffer', (done) => {
-      let node = dagNodes[1]
-      resolver.resolve(node, 'c/0/a/0/0/', (err, result) => {
-        expect(err).to.not.exist()
-        let trieNode = result.value
-        expect(result.remainderPath).to.eql('')
-        expect(isExternalLink(trieNode)).to.eql(false)
-        expect(Buffer.isBuffer(result.value)).to.eql(true)
-        done()
-      })
+    it('neck node resolves "c" down to buffer', () => {
+      const node = dagNodes[1]
+      const result = resolver.resolve(node, 'c/0/a/0/0/')
+      const trieNode = result.value
+      expect(result.remainderPath).to.eql('')
+      expect(CID.isCID(trieNode)).to.be.false()
+      expect(Buffer.isBuffer(result.value)).to.be.true()
     })
 
-    it('neck node resolves "b" down to branch', (done) => {
-      let node = dagNodes[1]
-      resolver.resolve(node, 'b/0/a/0/0/', (err, result) => {
-        expect(err).to.not.exist()
-        let trieNode = result.value
-        expect(result.remainderPath).to.eql('0/a/0/0/')
-        expect(isExternalLink(trieNode)).to.eql(true)
-        done()
-      })
+    it('neck node resolves "b" down to branch', () => {
+      const node = dagNodes[1]
+      const result = resolver.resolve(node, 'b/0/a/0/0/')
+      const trieNode = result.value
+      expect(result.remainderPath).to.eql('0/a/0/0')
+      expect(CID.isCID(trieNode)).to.be.true()
     })
 
-    it('neck node resolves "a" down to branch', (done) => {
-      let node = dagNodes[1]
-      resolver.resolve(node, 'a/0/a/0/0/', (err, result) => {
-        expect(err).to.not.exist()
-        let trieNode = result.value
-        expect(result.remainderPath).to.eql('0/a/0/0/')
-        expect(isExternalLink(trieNode)).to.eql(true)
-        done()
-      })
+    it('neck node resolves "a" down to branch', () => {
+      const node = dagNodes[1]
+      const result = resolver.resolve(node, 'a/0/a/0/0/')
+      const trieNode = result.value
+      expect(result.remainderPath).to.eql('0/a/0/0')
+      expect(CID.isCID(trieNode)).to.be.true()
     })
   })
 })
 
-function dumpTrieNonInlineNodes (trie, fullNodes, cb) {
-  trie._findDbNodes((nodeRef, node, key, next) => {
-    fullNodes.push(node)
-    next()
-  }, cb)
-}
-
-function contains (array, item) {
-  return array.indexOf(item) !== -1
+function dumpTrieNonInlineNodes (trie, fullNodes) {
+  return new Promise((resolve) => {
+    trie._findDbNodes((nodeRef, node, key, next) => {
```

Comment:
this should probably be promisified too, especially if the callback can return an error, in which case it's going to get turned into a return value via `resolve`

Reply:
I have no clue how to promisify this one. Can you help?

Reply:
Does `_findDbNodes()` resolve to something useful here or is the only purpose here to add the nodes to fullNodes?

Reply:
Regardless, I think it would be:

```javascript
return promisify((...args) => trie._findDbNodes(...args))((n, node, k, next) => {
   fullNodes.push(node)
   next()
})
```

Or, more readably:

```javascript
let _find = promisify((...args) => trie._findDbNodes(...args))
let _iter = (n, node, k, next) => {
   fullNodes.push(node)
   next()
}
return _find(_iter)
```

Reply:
Actually, to be more consistent with Rod’s earlier recommendation, you’d use `promisify(trie._findDbNodes.bind(trie))` instead of `promisify((...args) => trie._findDbNodes(...args))`.

Reply:
I've tried something similar, it doesn't work. What I now tried based on your suggestion is:

```js
async function dumpTrieNonInlineNodes (trie, fullNodes) {
  let _find  = promisify(trie._findDbNodes.bind(trie))
  let _iter = (nodeRef, node, key, next) => {
    fullNodes[nibbleToPath(key)] = node
    next()
  }
  return _find(_iter)
}
```

Reply:
Where did `fullNodes[nibbleToPath(key)]` come from?

Reply:
Also, `dumpTrieNonInlineNodes` doesn’t need to be an async function since it’s not using `await` anywhere and should be returning a promise in all the iterations of this so far.

Reply:
Oh, I see, there are two implementations of this function in tests but they are slightly different, the one this is replaced is using `fullNodes.push(node)` and the one in the other file is using `fullNodes[nibbleToPath(key)] = node`. Do you think that’s what is breaking it in this test? Looking at the diff this one should be `fullNodes.push(node)`.

Reply:
The currently working code:

```js
function dumpTrieNonInlineNodes (trie, fullNodes) {
  return new Promise((resolve) => {
    trie._findDbNodes((nodeRef, node, key, next) => {
      fullNodes[nibbleToPath(key)] = node
      next()
    }, resolve)
  })
}
```

The challenge is to promisify this somehow :)

Reply:
Try this and see if it still works.

```javascript
function dumpTrieNonInlineNodes (trie, fullNodes) {
  return new Promise((resolve, reject) => {
    trie._findDbNodes((nodeRef, node, key, next) => {
      fullNodes[nibbleToPath(key)] = node
      next()
    }, err => { if (err) return reject(err); resolve() })
  })
}
```

If it fails with that, then the reason it was succeeding is because you were swallowing the error.

If this still works then let’s just move on because we’ve handled the primary concern and it’s not like we’re worried about the perf benefits of promisify since this is just a test.

Reply:
Works, not with promisify though ;) Thanks!

Reply:
Very strange, they should do the same thing, but I’m not *that* surprised that an API that is using a one-off continuation passing function inside a function that’s a parameter to a standard callback is behaving oddly ;)

Reply:
ok, sounds messy, forget my pickiness then and just make it work

Reply:
No worries. In the other case promisify-ing it really made sense. I found it interesting to see that there are cases where strange callback-based APIs can't easily be promisified.